blackjax/optimizers/lbfgs.py

```diff
@@ -0,0 +1,289 @@
+    """
+    Minimize a function using L-BFGS
+
+    Parameters
+    ----------
+    fun:
+        function of the form f(x) where x is a flat ndarray and returns a real scalar.
+        The function should be composed of operations with vjp defined.
+    x0:
+        initial guess
+    maxiter:
+        maximum number of iterations
+    maxcor:
+        maximum number of metric corrections ("history size")
+    ftol:
+        terminates the minimization when `(f_k - f_{k+1}) < ftol`
+    gtol:
+        terminates the minimization when `|g_k|_norm < gtol`
+    maxls:
+        maximum number of line search steps (per iteration)
+
+    Returns
+    -------
+    Optimization results and optimization path
+    """
+
+    def lbfgs_one_step(carry, i):
+        (params, state), previous_history = carry
+
+        # this is to help optimization when using log-likelihoods, especially for float 32
+        # it resets stepsize of the line search algorithm back to stating value (INIT_STEP_SIZE) if
+        # it get stuck in very small values
+        state = state._replace(
+            stepsize=jnp.where(
+                state.stepsize < MIN_STEP_SIZE, INIT_STEP_SIZE, state.stepsize
+            )
+        )
+        # LBFGS use a rolling history, getting the correct index here.
+        last = (state.iter_num % maxcor + maxcor) % maxcor
+        next_params, next_state = solver.update(params, state)
+
+        # Recover alpha and update mask
+        s_l = next_state.s_history[last]
+        z_l = next_state.y_history[last]
+        alpha_lm1 = previous_history.alpha
+
+        alpha_l, mask_l = lbfgs_recover_alpha(alpha_lm1, s_l, z_l)
+
+        current_grad = previous_history.g + z_l
+        history = LBFGSHistory(
+            x=next_params,
+            f=next_state.value,
+            g=current_grad,
+            alpha=alpha_l,
+            update_mask=mask_l,
+        )
+        # check convergence
+        f_delta = (
+            jnp.abs(state.value - next_state.value)
+            / jnp.asarray([jnp.abs(state.value), jnp.abs(next_state.value), 1.0]).max()
+        )
+        not_converged = (next_state.error > gtol) & (f_delta > ftol) & (i < maxiter)
+        return (OptStep(params=next_params, state=next_state), history), not_converged
+
+    def non_op(carry, it):
+        return carry, False
+
+    def scan_body(tup, it):
+        carry, not_converged = tup
+        # When cond is met, we start doing no-ops.
+        next_tup = lax.cond(not_converged, lbfgs_one_step, non_op, carry, it)
+        return next_tup, next_tup[0][-1]
+
+    solver = jaxopt.LBFGS(fun=fun, maxiter=maxiter, maxls=maxls, history_size=maxcor)
+    value0, grad0 = jax.value_and_grad(fun)(x0)
+    state = solver.init_state(x0)
+
+    value0, grad0 = jax.value_and_grad(fun)(x0)
+    # LBFGS update overwirte value internally, here is to set the value for checking condition
+    state = state._replace(value=value0, stepsize=state.stepsize / 1.5)
```

Comment:
@miclegr any reason we reduce the initial step size here?

Reply:
No reason, I've left it there inadvertently while troubleshooting convergence issues

Reply:
Thanks! Resolving...